pkg/datapath/linux/probes/probes.go

```diff
@@ -487,3 +465,149 @@ func HaveV3ISA() error {
 	}
 	return nil
 }
+
+// CreateHeaderFiles creates C header files with macros indicating which BPF
+// features are available in the kernel.
+func CreateHeaderFiles(headerDir string, probes *FeatureProbes) error {
+	common, err := os.Create(filepath.Join(headerDir, "features.h"))
+	if err != nil {
+		return fmt.Errorf("could not create common features header file: %w", err)
+	}
+	defer common.Close()
+	if err := writeCommonHeader(common, probes); err != nil {
+		return fmt.Errorf("could not write common features header file: %w", err)
+	}
+
+	skb, err := os.Create(filepath.Join(headerDir, "features_skb.h"))
+	if err != nil {
+		return fmt.Errorf("could not create skb related features header file: %w", err)
+	}
+	defer skb.Close()
+	if err := writeSkbHeader(skb, probes); err != nil {
+		return fmt.Errorf("could not write skb related features header file: %w", err)
+	}
+
+	xdp, err := os.Create(filepath.Join(headerDir, "features_xdp.h"))
+	if err != nil {
+		return fmt.Errorf("could not create xdp related features header file: %w", err)
+	}
+	defer xdp.Close()
+	if err := writeXdpHeader(xdp, probes); err != nil {
+		return fmt.Errorf("could not write xdp related features header file: %w", err)
+	}
+
+	return nil
+}
+
+// ExecuteHeaderProbes probes the kernel for a specific set of BPF features
+// which are currently used to generate various feature macros for the datapath.
+// The probe results returned in FeatureProbes are then used in the respective
+// function that writes the actual C macro definitions.
+// Further needed probes should be added here.
```

Comment:
... but they also need their counterpart in `write*Header()`, don't they?

Would be nice to have both the writing patterns and the probes definition in a single place, but I understand it's not always a trivial 1:1 match (e.g. `BPF_HAVE_JIFFIES` we need to check for several program types) so not sure if easily doable.

Reply:
Yes, new macro strings will also have to be defined in the correct `write*Header()`. 
The reason to split probing and writing of macros was to be able to pass a "configurable" set of feature probes to `CreateHeaderFiles(...)` and it could potentially be called from a non-privileged context/test.

Reply:
Ok, I would simply extend the comment to point down to the `write*Header()` as well in that case, but fine by me otherwise

Reply:
But the test isn't calling `CreateHeaderFiles`, it's calling `writeFeatureHeader`.

Could the `FeatureProbes` consist of a list of conditions to check instead of a single boolean and then they would be resolved by checking their type (with the types you already defined) and calling the appropriate `HaveX` function for each?

Reply:
@pchaigno Would https://github.com/cilium/cilium/pull/21451#discussion_r1003200015 be an appropriate alternative to your suggestion?

I'd suggest we improve this in a follow-up PR either way.

Reply:
> Would https://github.com/cilium/cilium/pull/21451#discussion_r1003200015 be an appropriate alternative to your suggestion?

I don't think I understand how it avoids having two lists to update.

<br>

Here's what I have in mind to avoid the need to list the same features in two places:
```go
type feat interface {
	probe(probes *FeatureProbes)
	resolve(probes *FeatureProbes) bool
}

func (ph ProgramHelper) probe(probes *FeatureProbes) {
	probes.ProgramHelpers[ph] = (HaveProgramHelper(ph.Program, ph.Helper) == nil)
}

func (ph ProgramHelper) resolve(probes *FeatureProbes) bool {
	return probes.ProgramHelpers[ph]
}

func (m MapType) probe(probes *FeatureProbes) {
	probes.Maps[m] = (HaveMapType(m) == nil)
}

func (m MapType) resolve(probes *FeatureProbes) bool {
	return probes.Maps[m]
}

func main() {
	probes := FeatureProbes{
		Maps:           make(map[MapType]bool),
		ProgramHelpers: make(map[ProgramHelper]bool),
		Misc:           miscFeatures{},
	}
	features := map[string][]feat{
		"HAVE_LRU_HASH_MAP_TYPE": []feat{MapType{ebpf.LRUHash}},
		...
		"HAVE_JIFFIES": []feat{
			ProgramHelper{ebpf.CGroupSock, asm.FnJiffies64},
			ProgramHelper{ebpf.CGroupSockAddr, asm.FnJiffies64},
			ProgramHelper{ebpf.SchedCLS, asm.FnJiffies64},
			ProgramHelper{ebpf.XDP, asm.FnJiffies64},
		},
		...
	}

	// Gather the set of features currently listed in ExecuteHeaderProbes:
	var toProbe map[feat]bool
	for _, v := range features {
		for _, f := range v {
			toProbe[f] = true
		}
	}

	// Probe for those features:
	for f, _ := range toProbe {
		f.probe(&probes)
	}

	// From the new features map, compute a map[string]bool as listed in writeXHeader:
	for k, v := range features {
		supported := false
		for _, f := range v {
			supported = supported && f.resolve(&probes)
		}
		if supported {
			fmt.Printf("#define %s=1\n", k)
		}
	}
}
```

> I'd suggest we improve this in a follow-up PR either way.

Agreed. This shouldn't block this PR which is already an improvement.

Reply:
I see, there's some good ideas in there. Perhaps I wasn't clear enough in the reason behind the design choice. The goal is to end up with something along the lines of:

```go
_ := CreateHeaderFiles("bpf/testdata/", CommonFeatures{
  LRUHashMap: true,
  LPMTrieMap: false,
  ...
})
```

This will allow us to eventually drive the verifier tests from Go, and apply a technique similar to fuzzing to test all possible combinations of `#define`s that could potentally be enabled in the agent by a user. That's what I mean by separating the macro definitions (structs taken as input, which is just data) from the probe logic. This way, the macro writer can be used separately from the probes, which require permissions and will naturally return different output based on the underlying kernel.

Reply:
AFAICS, that is completely orthogonal to my idea. That is, you can do what you're suggesting using my approach, by initializing `probes` with those test values instead of using `f.probe()`. It's the whole reason I separated `feat.probe()` from `feat.resolve()`.